### ISSUE TITLE
Inconsistency in absolute position index across clients after an undo operation

### ISSUE THREAD

**Issue by nimeshnayaju** [yjs/yjs#638]
**Checklist**

* [x] Are you reporting a bug? Use github issues for bug reports and feature requests. For general questions, please use https://discuss.yjs.dev/
* [x] Try to report your issue in the correct repository. Yjs consists of many modules. When in doubt, report it to https://github.com/yjs/yjs/issues/

**Describe the bug**
Absolute position index after an undo operation becomes inconsistent across different clients in a collaborative text editing session. I am mostly curios if this expected behavior. If so, this would make use of `RelativePosition` for commenting in a collaborative text document difficult to implement or almost unusable.

**To Reproduce**

See reproduction here: [Stackblitz Link](https://stackblitz.com/edit/vitest-dev-vitest-vc9ngf?file=test%2Fbasic.test.ts,test%2Fyjs.test.ts%3AL11-L11&initialPath=__vitest__/)

Steps:
1. Two clients (Client A and Client B) start with synchronized documents with a shared text type under key `text`. Changes made in Client A are applied to Client B.
2. Client A inserts `abcde` to the shared `text`.
3. Create a relative position at index 2 of the shared `text`.
4. Convert this relative position to absolute positions on both Y.js Doc on Client A and Client B, and verify that they point to the same position.
5. Delete the character at index 2 (`c`) from the shared `text` and verify that the changes (deletion) are reflected on both clients.
7. Perform an undo operation using UndoManager to revert the deletion.
8. Convert the same initial relative position again to absolute positions on both Client A and B.

**Expected behavior**
After undoing the deletion, both documents should revert to 'abcde' and the absolute position at index 4 should be consistent across both clients.

**Actual behavior**
The text reverts correctly, but the absolute position points to index 2 in Client A and incorrectly to index 3 in Client B.

**Environment Information**
 - Yjs v3.6.14

**Comment by dmonad** [comment]
Yes, this is the expected behavior. I'll try to explain what is happening internally so you can understand it better. Below is a fix to get consistent behavior.

In Yjs we can't undo deletions (as in undeleting the internal item representation). Instead, we re-insert content. The advantage is that we can perform very aggressive garbage-collection. This means that after performing undo/redo, relative positions are not always placed at the position where they intuitively should be. This is a tradeoff - but I think it's worth it.

Intuitively, the relative position should be at index 2 after performing undo. We can't ensure that.

When you insert `ab|cde`, and point the relative position to the character `c`, the relative position will still point to the deleted character `c`, even though it is not visible anymore. After undoing the deletion, a new character `c` will be inserted: `abc|cde` (note that the second character `c` is basically invisible now). 

The correct relative position in your scenario (at least technically) is position 3.

Now, to the weird part. We can restore the intuitively correct position if we have access to the UndoManager that performed the undo operation. The client that performed undo knows more about the document state. It knows that the second `c` character is "redone" by the first `c` insertion. So it "follows" the "redo-path" and figures out that position 2 is more accurate. However, the "redo-path" is not persisted in the Yjs document and will eventually be lost.

Since relative positions are mainly used to restore the cursor state after a client performs an undo, this is the default behavior. In your case, I recommend not following the redo path to achieve consistent behavior.

TLDR:
Add `false` to `createAbsolutePositionFromRelativePosition(rpos, doc, false)` to get consistent behavior across all clients. This disables "following the redo-path". This feature is available in the latest Yjs release.

Feel free to ping me to chat more about a commenting solution based on relative positions.

**Comment by JuanM04** [comment]
@dmonad awesome explanation about the internals of the library!

I'm facing the same issues mentioned here. After upgrading to the latest version and applying the fix you commented, the consistency issues disappeared.

Although, a different problem appeared. Now, it seems like the creation of relative positions are broken. If you add `false` to the very first `createAbsolutePositionFromRelativePosition` in the minimal reproduction (line 32), you can see that the relative position seems to point to 0 instead of 2.

```js
// Verify that the absolute position on 'docA' and 'docB' are the same
const absolutePositionOnDocABeforeDeletion =
  createAbsolutePositionFromRelativePosition(relativePosition, docA, false);
expect(absolutePositionOnDocABeforeDeletion?.type).toBe(textA);
expect(absolutePositionOnDocABeforeDeletion?.index).toBe(2);
```

Am I missing something? Because I can see the same issues in my app. When I remove that last `false`, it seems to work fine (until I do some undo/redo, of course).

**Comment by dmonad** [comment]
The UndoManager won't restore the relative position. Technically, text that is "redone" (recreated after it has been undone) is new content. The relative position still points to the old content.

Either you make the comment part of the document by using marks instead (which might be undesirable if you want read-only users to be able to make comments) or you need to write some extra code to adapt comment ranges after you hit undo.

After hitting undo, you could check if the relative position needs to be updated (`createAbsolutePositionFromRelativePosition(relativePosition, docA, false) !== createAbsolutePositionFromRelativePosition(relativePosition, docA, true)` ↦ position needs to be updated to `createAbsolutePositionFromRelativePosition(relativePosition, docA, true);`)

Also, you might want to handle the case where a user splits a paragraph. If that happens, you probably want to update the relative position as well, because the end of the range will point to the deleted text, not the newly inserted text.

**Comment by JuanM04** [comment]
Correct me if I'm wrong, but the issue I'm highlighting isn't affected by an UndoManager. See [this reproductions](https://stackblitz.com/edit/vitest-dev-vitest-gifybb?file=test%2Fyjs.test.ts), forked from the previous one but with the UndoManager removed.

See how those `createAbsolutePositionFromRelativePosition` result in different positions?
